CRPS and SCRPS

LeeviLindgren opened the pull request:
Addresses #136. Adds functions to compute `crps` and `scrps` and their `loo` versions. Happy to hear feedback!

avehtari commented on R/crps.R:
```diff
@@ -0,0 +1,153 @@
+#'
+#' To compute (S)CRPS user needs to provide two sets of draws from the
+#' predictive distribution. I.e. one needs to call, for instance,
+#' `posterior_predict()` two times and pass the returned values on to crps
```
crps -> CRPS or `crps` ?

yao-yl commented:
Some math:


###  CRPS score
The  CRPS  for forecast CDF $F$ is 
$$\mathrm{crps}(F,y)=E_X|X-y|- \frac{1}{2}E_{X,X^\prime}|X-X^\prime|.$$
where $X$ and $X'$ are two independent copies from $F$.

Using these predictive draws  $z_{m1},  \dots, z_{mS}$, we can compute the crps of the model on the $m$-th datapoint by 

$$ \widehat {\mathrm{crps}}_{m}= \widehat {\mathrm{crps}}_(z_{m1}, \dots, z_{ms}, z_{m})= \frac{1}{S} \sum_{s=1}^S  |z_{ms}-z_{m}| -
\frac{1}{2S^2} \sum_{s, s'=1}^S |z_{ms}- z_{ms'}|.$$
 
 
 
 
 
### Fast LOO approximation

For each draw $\theta_s$  from posterior $p(\theta \vert   y)$, we simulate a prediction 
 $$\tilde y_{1s}, \tilde y_{2s}, \dots,  \tilde y_{ns}$$ 
from the pointwise sampling distribution $p(y \vert  \theta)$.

 
Fix a data point $m$, we need to compute the  LOO-IS ratios and ``product LOO-IS ratio''


$$r_{s} =    \frac{ p^{-1}(y_m | \theta_{s}    )}      {\sum_{j=1}^S          p^{-1}(y_{m}  | \theta_{ j})    }  $$
 
$$ r_{ss^\prime} =     \frac{ p^{-1}(y_{m}  | \theta_{s}    ) p^{-1}(y_{m}  | \theta_{s^{\prime}}    )}      {\sum_{j, j^\prime=1}^{S}          p^{-1}(y_{m} | \theta_{ j})    p^{-1}(y_{m}  | \theta_{j^{\prime}}    )}   $$


We estimate the pointwise crps score by 


$$
   \widehat {\mathrm{crps}}_{m}^{\mathrm{loo}}   =   \sum_{s}    r_{s}    \vert   \tilde y_{ms}-y_{m} \vert  -
	\frac{1}{2} \sum_{s, s'=1}^S   r_{ss^\prime} \vert   \tilde y_{ms}- \tilde y_{ms'} \vert    
$$

The second term needs to run PSIS on the product ratios.


 ### Jensen inequality  
These two schemes are different:
$$E_{\theta} E_{ x_1,  x_2 \vert \theta } {  abs( x_1 - x_2  )     } \neq E_{\theta_1, \theta_2} E_{ x_1  \vert \theta_1   x_2 \vert \theta_2 } {  abs( x_1 - x_2  )     }$$

LeeviLindgren commented:
Thanks @yao-yl! I compared the `crps` with the old and suggested version to the one from [scoringRules](https://github.com/FK83/scoringRules) `crps_sample`-function.  With the old implementation, the correlation is close to 1 but there is a small bias (~0.006). When the shuffling is added, estimates are drastically different:
![image](https://user-images.githubusercontent.com/48201042/176362375-31cd8b33-293e-409c-8200-1cfe913976ad.png)
It is not yet clear to me why this happens.
```
library(rstanarm)
library(ggplot2)

crps_srules <- numeric()
crps_custom <- numeric()

for(i in 1:100) {
  if(i %% 10 == 0) print(i)
  x <- rnorm(100)
  y <- rnorm(1) + x * rnorm(1) + rnorm(100)
  fit <- stan_glm(y ~ x, data = data.frame(y, x), refresh=0)

  crps_srules[i] <- mean(-scoringRules::crps_sample(y, t(posterior_predict(fit))))
  crps_custom[i] <- crps(posterior_predict(fit), posterior_predict(fit), y)$estimate[1]
}
cor(crps_srules, crps_custom)
qplot(crps_srules, crps_custom) + geom_abline()
```

yao-yl commented:
@LeeviLindgren the separation for the matrix and vector looks good.
One more question: when computing Exx, in theory we can average over all permutations to reduce variance (In the current version, we only compare one possible permutations x1, and x2[shuffle]). 
If we do generate more permutations, the variance will be reduced for Exx, while Exy has a fixed variance term. On the other hand, the estimate of Exx could have a higher variance to begin with because it involves two random copies.

Would it be a good idea to add this permutation times as the function input (at this moment it is 1):  computing EXX  multiple times and taking the average in crps and loo_crps. It is a further research topic to how to choose the reasonable number of permutations to balance the variance of Exx and Exy.

jgabry commented:
> @LeeviLindgren the separation for the matrix and vector looks good. One more question: when computing Exx, in theory we can average over all permutations to reduce variance (In the current version, we only compare one possible permutations x1, and x2[shuffle]). If we do generate more permutations, the variance will be reduced for Exx, while Exy has a fixed variance term. On the other hand, the estimate of Exx could have a higher variance to begin with because it involves two random copies.
> 
> Would it be a good idea to add this permutation times as the function input (at this moment it is 1): computing EXX multiple times and taking the average in crps and loo_crps. It is a further research topic to how to choose the reasonable number of permutations to balance the variance of Exx and Exy.

Did these questions ever get resolved? 

Trying to figure out if this is ready or if we're still waiting on some changes.

LeeviLindgren commented:
Sorry for the late reply! @jgabry I implemented the permutation option suggested by @yao-yl. I did some checks today and fixed a minor bug and from my side, it is good for reviewing.

jgabry commented:
I'm finally getting ready to prepare another loo release in order to fix a test failure for CRAN, but we can also include this PR. @LeeviLindgren the changes look good to me, thanks! I have a few small changes to suggest for the documentation, but I can do those after merging this. 

@avehtari Are you good with this? If so, I think we're ready to merge it.

jgabry commented:
> I'm finally getting ready to prepare another loo release in order to fix a test failure for CRAN, but we can also include this PR. @LeeviLindgren the changes look good to me, thanks! I have a few small changes to suggest for the documentation, but I can do those after merging this.
> 
> @avehtari Are you good with this? If so, I think we're ready to merge it.

Also @LeeviLindgren could you add me as a collaborator in your fork of the loo repository? I thought it would let me push to this PR but it's not letting me. I guess you might need to add me as a collaborator in your repository settings.

jgabry commented:
Great, thanks! I might make a few small changes before merging, but in general this PR looks good. Thanks again @LeeviLindgren. I will also add you to the package DESCRIPTION file as a contributor to the package.
